### DISCUSSION TITLE
Other Python declarative native GUI projects

### DISCUSSION THREAD

**Discussion by jamesdbrock** [pyedifice/pyedifice#43] [General]
## RE-WX

https://github.com/chriskiehl/re-wx

> re-wx is a library for building modern declarative desktop applications. It's built as a management layer on top of WXPython

## PUI

https://github.com/buganini/PUI

> PUI is a declarative UI framework with two-way data binding

## Reacton

https://github.com/widgetti/reacton

> A way to write reusable components in a React-like way, to make Python-based UI's using the ipywidgets ecosystem

https://maartenbreddels.medium.com/advance-your-ipywidget-app-development-with-reacton-6734a5607d69

> The `use_effect` and `use_state` functions in Reacton are called “hooks”

Great API design, but sadly not native — only for Jupyter iPyWidgets.

## Collagraph 

https://github.com/fork-tongue/collagraph

> Python interfaces in a declarative manner with plain render functions, component classes or even single-file components using Vue-like syntax

## pyreaqtive

https://github.com/polfeliu/pyreaqtive

> pyreaqtive is a framework for developing applications in pyqt/pyside with reactive programming

## Alfort

https://github.com/ar90n/alfort

> simple and plain ELM-like interactive applicaiton framework for Python

## Flet

https://github.com/flet-dev/flet

> With Flet you just write a monolith stateful app in Python only and get multi-user, realtime Single-Page Application (SPA).

Flet seems promising, because Flet is based on Flutter, and Flutter is declarative. But Flet made some bad architectural choices, and is imperative. https://github.com/flet-dev/flet/issues/55#issuecomment-1600194986

## Kivy

https://github.com/kivy/kivy

https://kivy.org/doc/stable/guide/lang.html#concept-behind-the-language

> the KivyMD library allows you to write Python code in a declarative style. Just as it is implemented in Flutter/Jetpack Compose/SwiftUI.

> Declarative programming style in Python code in the KivyMD library is an experimental feature. Therefore, if you receive errors, do not hesitate to create new issue in the KivyMD repository.

https://kivymd.readthedocs.io/en/1.0.1/behaviors/declarative/index.html#declarative-style-with-python-code

## Castella

https://github.com/i2y/castella

https://speakerdeck.com/i2y/a-cross-platform-pure-python-declarative-ui-framework

## Enaml

https://github.com/nucleic/enaml

> * A declarative programming language, with a Pythonic flavour.
> * Dozens of widgets, ready to go out-of-the-box (built on Qt).

## DearPyGui

https://github.com/hoffstadt/DearPyGui

> it uses the immediate mode paradigm

https://github.com/nix-community/nur-combined/blob/master/repos/some-pkgs/pkgs/dearpygui/default.nix

## Python Concur

https://github.com/potocpav/python-concur

> Concur is a Python GUI framework based on synchronous generators.
>
> It is a port of [Concur for Purescript](https://github.com/ajnsit/purescript-concur), implemented on top of the [Dear ImGui](https://github.com/ocornut/imgui) C++ UI library.

## LayoutX

https://github.com/Bomberus/LayoutX

> Declarative tkinter UI with reactive data binding

## PynamicUI

https://github.com/zacharie410/PynamicUI

> create dynamic web-like UIs using a declarative syntax
> PynamicUI requires the CustomTkinter library.

## Slint-python

https://github.com/slint-ui/slint/tree/master/api/python

> [Slint](https://slint.dev/) is a UI toolkit that supports different programming languages. Slint-python is the integration with Python.

> “P.S. The founders are ex-Qt (with 16+ years of working for Qt) .. and we decided to develop a 'better QML' (if you may say) based on our experiences.” https://www.reddit.com/r/Python/comments/18mdpig/comment/ke5nulu/

https://nlnet.nl/project/PythonicSlint/

https://pypi.org/project/slint/

## reactpy

https://github.com/reactive-python/reactpy

Renders to HTML.

> [“This is cute, but it would be more helpful to have react-style GUI programming in python, with Qt (Quick or widgets) as the backend (instead of DOM).”](https://news.ycombinator.com/item?id=36220342)

**Comment by jamesdbrock** [comment]
Hi @buganini and @berendkleinhaneveld !

There are too many of these projects. Let’s try to combine our efforts? What do you think about Edifice?

**Reply by buganini** [reply]
Just updated PUI's description as follow:
Python Declarative UI Framework. PUI doesn't do UI itself, it turns imperative UI libraries into declarative flavor with virtual DOM and aims to maintain interoperability.

**Reply by pushfoo** [reply]
First, thank you for compiling this list.

> There are too many of these projects.

Not all of these projects have the same goals or focus as Edifice. Using PySide / PyQt implies it's a desktop oriented library. The best way to compete with others in the space would be to address their pain points. Using only Python syntax already eliminates many of them.

My questions / known pain points:

1. Is there a way to integrate it with a GL context? The goal would be:
    1. Use Edifice to render UI
    2. Allow a GL binding to render game / model details
2. Will Edifice avoid a CLA like the one Beeware's Toga uses?

> Python Declarative UI Framework. PUI doesn't do UI itself, it turns imperative UI libraries into declarative flavor with virtual DOM and aims to maintain interoperability.

This is fascinating. Many pyglet contributors (myself included) want to improve its GUI system. I've also wanted to make a generic layout helper library for use in both GUI and game world generation. PUI seems like it could be useful for wrapping these.

My current questions:

1. Does PUI provide any abstractions for rendering backends which use different coordinate systems? (origin at top left vs OpenGL's default bottom left)
4. Is there benefit / use for PUI in Edifice?

> Kivy

As a mobile-first framework, it doesn't address the same problems as Edifice. The main similarities seem to be:

* declarative structures
* the capacity to bind functions on them

When I last used Kivy, its widget set felt out of place on the desktop. Years-old desktop issues remain open (example: https://github.com/kivy/kivy/issues/6708), suggesting desktop still isn't a priority.

**Reply by berendkleinhaneveld** [reply]
Edifice looks interesting; certainly the dev tools (reloading and the inspector) are a nice touch!

A few things that I need/like which seem to be missing here (and why Collagraph exists) are as follows:
* Render to different frameworks: Qt for windowing/UI, Pygfx for 3D/2D visualizations
* No (significant) abstraction over Qt API, in order to keep fine-grained control
* Vue-like reactivity system, which should have some benefits for deciding what (not) to (re-)render
* Vue-like template syntax, as we have some people in our team which are familiar with that. This is not a necessity but makes for improved readability of 'elaborate' components.

On my 'I-can-dream' wish list would be: having no VDOM, more like Svelte for instance. As Python is already slower than JS, it seems to me that having a VDOM is actually not that great. Who knows, something for a future version of one of these libraries?

Thanks for the list, it's quite interesting to see all the efforts and compare them a bit. I'm getting a little bit inspired to design a kind of template syntax, but within Python syntax: I kinda like what PUI is doing.

**Reply by jamesdbrock** [reply]
> Edifice looks interesting; A few things that I need/like which seem to be missing here (and why Collagraph exists) are as follows:
>
> - Vue-like reactivity system, which should have some benefits for deciding what (not) to (re-)render

Oh yes. Edifice is terrible at deciding what to re-render. https://github.com/pyedifice/pyedifice/issues/38

> - Vue-like template syntax

I’m opposed to any special template syntax. Multi-language interop causes enormous problems and should only be a last resort. If Python is not expressive enough to declare a widget tree, then what are we even doing with Python?

> design a kind of template syntax, but within Python syntax: I kinda like what PUI is doing.

Yeah now we're talking. I love the `with` context manager syntax that [PUI](https://github.com/buganini/PUI#state--data-binding) and [Reacton](https://reacton.solara.dev/en/latest/api/#hooks) are using to declare widget trees. Best Python syntax I've ever seen.

**Reply by berendkleinhaneveld** [reply]
> then what are we even doing with Python?

Making use of its rich ecosystem :) I get your point though. Collagraph's python syntax could certainly be improved, maybe even up to a point where the vue-like template syntax can be left behind. Keeping everything within Python syntax makes it a lot easier to benefit from all the tooling that is available, like auto-formatters, linting, etc.

Think I'm going to try and see how (and if) all features from a Vue-template would translate to a PUI/Reacton style syntax.

**Reply by berendkleinhaneveld** [reply]
I gave the `with` syntax some thought, but it seems that there are some limitations to this approach.

In frameworks like Vue and Svelte, the template is compiled to a function that can generate the VDOM. During this compilation step a lot of optimisations can be applied so that for instance static elements and properties will be excluded from any tree-diffing that has to be performed.

The `with` syntax doesn't really allow for that: it just builds the VDOM at a certain 'moment-in-time' and then the full VDOM will have to be diffed with the previous VDOM. In order to be able to apply optimisations you would need to create an AST of the 'content' function, and then try and analyse that. However, because Python's syntax is so expressive, it becomes impossible to do that. So I'm a little bit less enthusiastic about this approach, but I'm not completely giving up hope yet...

BTW: if this is not the appropriate place to share these thoughts, please let me know :)

**Reply by jamesdbrock** [reply]
> The `with` syntax doesn't really allow for that: 
> 
> BTW: if this is not the appropriate place to share these thoughts

Thanks for your thoughts, that's very interesting. (This is *exactly* the kind of conversation that I want to have here.)

**Reply by buganini** [reply]
About the VDOM diffing, To prevent diffing a huge VDOM, PUI has a subview design and my worst thought is to construct and diff VDOM in C++ ;P

**Reply by berendkleinhaneveld** [reply]
I'm interested in how that subview design works, will take a look at it soon.
I've also been thinking about diffing in C/C++/rust, but I'm afraid that there is going to be way too much overhead from going in and out of python context for it to provide any real benefits... 
I've been reading about svelte 5: https://svelte.dev/blog/runes and it links to some interesting articles, also about SolidJS.  https://dev.to/ryansolid/a-hands-on-introduction-to-fine-grained-reactivity-3ndf -> in Collagraph we use observ, which is basically a signal lib.

What would be really cool is if we can have a Python project which uses the concepts from Solid. Their JSX template syntax has some tricks which might actually be translatable to a Python version with the `with` construct.

**Reply by maartenbreddels** [reply]
I don't think 'DOM' diffing needs to be a problem, as long as you by default assume purity (like React's pure components using memo), and only diff what is needed. Too large components should be avoided by a user, maybe Reacton components are like PUI's subview?.

In solara (which is built on top of Reacton), we introduce reactive variables, which are similar to signals in preact:
https://solara.dev/api/reactive

```python
import solara

counter = solara.reactive(0)

def increment():
    counter.value += 1


@solara.component
def CounterDisplay():
    solara.Info(f"Counter: {counter.value}")


@solara.component
def IncrementButton():
    solara.Button("Increment", on_click=increment)


@solara.component
def Page():
    IncrementButton()
    CounterDisplay()
```

Where components automatically subscribe to update (i.e. automatically rerender) when the state of a reactive variable changes.

**Reply by pushfoo** [reply]
I have an update on these earlier points:

> Is there a way to integrate it with a GL context?

> Render to different frameworks: Qt for windowing/UI, Pygfx for 3D/2D visualizations

The following suggest 3D integration is definitely possible here as well:

1. Elm has (somewhat outdated) [WebGL examples](https://github.com/elm-community/elm-webgl/tree/master/examples)
2. I've been updating a [pyglet example integrating PySide2/PyQt5-based](https://github.com/pyglet/pyglet/blob/master/contrib/qt_sprite_preview.py) to familiarize myself with Python's Qt bindings

The `QOpenGLWidget` the example uses also exists in [PySide6](https://doc.qt.io/qtforpython-6/PySide6/QtOpenGLWidgets/QOpenGLWidget.html). However, there also seems to be a [`PySide6.Qt3DCore`](https://doc.qt.io/qtforpython-6/PySide6/Qt3DCore/index.html) module, which provides more abstractions than pyglet does at the moment.

**Comment by jamesdbrock** [comment]
Hi @maartenbreddels , I love your Reacton project. I love your choice to call components using a [`with`](https://docs.python.org/3/reference/compound_stmts.html#with) context manager. And congratulations for implementing [Hooks](https://reacton.solara.dev/en/latest/api/#hooks) in Python; I believe you’re the first person to do that.

**Reply by maartenbreddels** [reply]
Thank you @jamesdbrock !
Let's continue at https://github.com/widgetti/reacton/issues/20

**Comment by jamesdbrock** [comment]
Edifice now has

1.  `with`-statement [tree declaration syntax](https://pyedifice.github.io/tutorial.html)
2. [Hooks](https://pyedifice.github.io/hooks.html)
